### PR TITLE
[Feature] A calibration-free RTN-based quantization for accurate and accelerated INT4/INT8 inference

### PR DESCRIPTION
**Pull request by sakogan**
This PR adds basic support for RTN quantization, as a first step for supporting a calibration-free RTN-based quantization for accurate and accelerated INT4/INT8 inference (see [this paper](https://arxiv.org/abs/2505.15909) for details).

RTN is a simple quantization method that does not require any calibration data nor a corresponding calibration process.
As such, it can be applied on-the-fly (i.e., while loading an original model) in a fast and cheap way, even on a system that does not have enough memory to host the original (unquantized) model. Yet, RTN is often believed to lag behind more advanced quantization techniques in two crucial areas – generation throughput and accuracy.

As this [paper](https://arxiv.org/abs/2505.15909) shows, both issues can be alleviated, through the use of efficient CUDA kernels based on [Marlin](https://github.com/IST-DASLab/marlin) (for throughput) and selective quantization (for accuracy). The latter is a simple mechanism that allows a user to select layers and/or specific linear modules that should be quantized to a higher precision. For instance, leaving just a part of one layer of Llama-3.1 70B model in 8 bit precision, while quantizing the rest of that layer and all other 79 layers into 4 bits leads to a substantially improved recovery rate, on-par with or better than other techniques:
Screenshot 2025-05-27 at 12 38 22 PM
Note that this adds less than 0.05 bits per weight on average, resulting in only insignificant memory increase.

As noted above, this PR is for basic Python-based implementation for RTN that supports quantizing models on-the-fly.
Once approved, we intend to enhance it with:

- Optimized CUDA (Marlin-based) kernels (for fast GEMM operations).
- Support for selective quantization (for improved accuracy)
- Support for MoE models

### CODE REVIEW

**Comment by sakogan** [comment]
@mgoin @robertgshaw2-redhat @tlrmchlsmth Can you, please, take a look at this PR and let me know if you have any comments? Thanks!

File: vllm/model_executor/layers/quantization/rtn.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    def apply(self,
+              layer: torch.nn.Module,
+              x: torch.Tensor,
+              bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+        qweight = layer.weight
+        scale = layer.scale
+
+        weight = rtn_dequantize(qweight, scale)
+        out = F.linear(x, weight)
+
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/rtn.py]: Just to make it explicit to torch
```suggestion
        weight = rtn_dequantize(qweight, scale)
        out = F.linear(x, weight)
        del weight
```

File: tests/quantization/test_rtn.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright © 2025, Oracle and/or its affiliates.
+"""Tests RTN quantization startup and generation, 
+doesn't test correctness
+"""
+import pytest
+
+from tests.quantization.utils import is_quant_method_supported
+
+MODELS = ["microsoft/Phi-3-mini-4k-instruct"]
```
Comment (mgoin) [inline tests/quantization/test_rtn.py:10]: Is there a reason why you are using Phi here, like to get around sharded weight loading? IIRC Phi models have their mergable layers like q/k/v already merged in the checkpoint as `qkv_proj`. I notice you override the weight loading with your RTNParameter class so I'm curious if it works with an un-merged checkpoint like Llama

Comment (sakogan) [reply tests/quantization/test_rtn.py:10]: yes, absolutely it works with any dense model, including un-merged LLama checkpoints. The Phi model is an arbitrary choice of a small dense model, happy to change it to something else

**Review by mgoin — APPROVED** [review]

**Comment by renjie0** [comment]
"Yet, RTN is often believed to lag behind more advanced quantization techniques in two crucial areas – generation throughput and accuracy." How does it look like now with your latest improvement?

**Comment by YouNeedCryDear** [comment]
> "Yet, RTN is often believed to lag behind more advanced quantization techniques in two crucial areas – generation throughput and accuracy." How does it look like now with your latest improvement?

You can take a look at the paper mentioned above for more detail.
